src/v/cloud_storage/remote.cc

```diff
@@ -325,6 +325,21 @@ ss::future<upload_result> remote::upload_segment(
             co_return upload_result::cancelled;
         }
 
+        // If the uploading partition wishes to stop while an upload is in
+        // progress, enable this by forcefully shutting down the HTTP client.
+        auto as_sub = fib.root_abort_source().subscribe(
+          // Lifetimes:
+          // - `lease` is scoped to this function, as is the
+          // abort source subscription: as will always be deregistered
+          // before lease is destroyed.
+          // - `ctxlog` is also function scoped.
+          [&lease, &ctxlog]() noexcept {
+              vlog(
+                ctxlog.debug,
+                "Cancelling in-flight requests on partition shutdown");
+              lease.client->shutdown();
```

Comment:
This PR didn't cover all the places we call into a client, but yeah, if we come back and make all uses of clients abort-safe, we should in principle no longer need the early shutdown.  To make that really robust against future mistakes, it would be neat to make the client lease itself carry the abort source, so that it's impossible to get a client without providing an abort source.